src/main/java/org/elasticsearch/index/translog/fs/FsTranslog.java

```diff
@@ -359,70 +431,81 @@ public Location add(Operation operation) throws TranslogException {
                 current.sync();
             }
 
-            assert new BytesArray(current.read(location)).equals(bytes);
-
-            FsTranslogFile trans = this.trans;
-            if (trans != null) {
-                try {
-                    location = trans.add(bytes);
-                } catch (ClosedChannelException e) {
-                    // ignore
-                }
-            }
-            Releasables.close(bytes);
-            released = true;
+            assert current.assertBytesAtLocation(location, bytes);
             return location;
         } catch (Throwable e) {
             throw new TranslogException(shardId, "Failed to write operation [" + operation + "]", e);
         } finally {
-            rwl.readLock().unlock();
-            if (!released && out != null) {
+            if (out != null) {
                 Releasables.close(out.bytes());
             }
         }
     }
 
     @Override
-    public FsChannelSnapshot snapshot() throws TranslogException {
-        while (true) {
-            FsChannelSnapshot snapshot = current.snapshot();
-            if (snapshot != null) {
-                return snapshot;
+    public Snapshot snapshot() {
```

Comment:
you should definitely add javadocs to this, all the possible off-by-one errors if someone misinterprets what this is doing (most likely to be me, lol)